proposals/0000-swift-settings.md

```diff
@@ -0,0 +1,446 @@
+specific target. This would allow for SwiftPM users to take advantage of the
+discoverability provided by specifying compiler flags using strongly typed APIs
+instead of the stringly typed APIs that `SwiftSetting` uses today. This
+generalized API also would ensure that `SwiftSetting` does not need to be
+updated to support more kinds of command line options since the compiler would
+be able to "inject" support for such options into `SwiftPM`. This would require
+us to allow for `SwiftSettings` structs to be constructed and provide state
+within them that `SwiftPM` can process to determine the relevant behavior. Since
+`SwiftSetting` is a resilient struct, we can in the future add this support
+without breaking ABI or API.
+
+NOTE: Since the `SwiftPM` API would specify in its signature that it takes a
+`Swift.SwiftSetting`, the user will still be able to take advantage of the type
+checker to avoid needing to prefix any calls to static methods on
+`Swift.SwiftSetting`.
+
+## Alternatives considered
+
+### Using a bespoke syntax instead of a macro
+
+Instead of using a macro, we considered using a bespoke syntax called
+`compilerSettings` that looked as follows:
+
+```swift
+compilerSettings [
+  .warningsAsErrors(.concurrency),
+  .strictConcurrency(.complete),
+  .defaultIsolation(MainActor.self)
+]
+```
+
+We decided not to go with this approach since:
+
+* We would be introducing an additional unneeded bespoke syntax to the language.
+* Macros are already recognized by users as a manner to change compiler behavior
+  at compile time.
+
+### Reusing `SwiftSetting` from SwiftPM
+
+Instead of using our own `SwiftSetting` struct, we could reuse parts of
+`SwiftPM`. The authors upon investigating this approach noticed that If one
+attempted to move SwiftPM's `SwiftSetting` into the standard library, one would
+have to expose many parts of SwiftPM's internal types into the standard library
+such as `BuildSettingData`, `BuildSettingCondition`, `Platform`,
+`BuildConfiguration`, and more.
+
+### Using a "fake" syntatic macro instead of a real macro
+
+Instead of using a real macro, we could use a bespoke "fake" macro. This would
+involve the compiler providing custom parsing for `#SwiftSettings`. The compiler
+would then explicitly pattern match against the syntax for the various command
+line options so the struct `SwiftSetting` would not be required.
+
+The reason why we decided not to go with this approach is that:
+
+1. It prevents the ability to perform lookup at point in the IDE and introduces
+   more bespoke code into the compiler to support this feature.
+2. It would prevent the introduction of new SwiftPM APIs in the future that take
+   a `Swift.SwiftSetting`.
+
+### "import feature" syntax
+
+We could also use an "import" syntax to enable features like Python and
+Scala. We decided to not follow this approach since `#` is a manner in swift of
+saying "compile time".
```

Comment:
Further, we'd be introducing a third way of specifying build settings (command line spelling, package manifest spelling, and an import-like spelling). The approach in this proposal was taken primarily to keep a consistent way of spelling build settings in source code.